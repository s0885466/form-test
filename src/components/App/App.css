.errors{
    font-size: 0.5rem;
    color:red;
}
label {
    display: block;
}

input.error {
    border: 2px solid red;
}