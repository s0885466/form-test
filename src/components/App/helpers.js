export function validateName(name) {
    return name.length < 6
        ? 'Name must be more than 6 symbols'
        : ''
}

export function validateEmail(email) {
    const pattern = /^(([^<>()[\]\\.,;:\s@\"]+(\.[^<>()[\]\\.,;:\s@\"]+)*)|(\".+\"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
    return pattern.test(email) ? '' : 'Email is invalid';
}

export function validatePassword(password) {
    return password.length < 6
        ? 'Password must be more than 6 symbols'
        : ''
}