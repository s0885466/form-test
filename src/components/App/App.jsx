
import React, {Component} from 'react';
import './App.css';
import {validateEmail, validateName, validatePassword} from "./helpers";
import RenderInput from "../render_input/RenderInput";
class App extends Component<{}> {
    state = {
        values: {
            name: '',
            email: '',
            password: '',
        },
        errors: {
            name: '',
            email: '',
            password: '',
        },

        validators: {
            name: validateName,
            email: validateEmail,
            password: validatePassword
        },
        isSubmitDisables: true
    };

    submit = e => {
        e.preventDefault();
    };

    onChange = (type:string) => (e:KeyboardEvent) => {
        const {value} = e.target;
        const values = {...this.state.values};
        values[type] = value;
        this.onError(value, type);
        this.setState({values}, () => {
            this.setSubmitDisabled();
        });
    };

    onError = (value, type) => {
        const error = this.state.validators[type](value);
        const errors = {...this.state.errors};
        errors[type] = error;
        this.setState({errors});
    };

    setSubmitDisabled = () => {
        const {values, validators} = this.state;
        for (let key in values) {
            if (validators[key](values[key])) {
                this.setState({isSubmitDisables: true});
                return;
            }
        }
        this.setState({isSubmitDisables: false});
    };

    render() {
        const {values, errors, isSubmitDisables} = this.state;
        return (
            <div>
                <form onSubmit={this.submit}>
                    <RenderInput
                        name={'name'}
                        label={'Name'}
                        placeholder={'type name'}
                        value={values.name}
                        onChange={this.onChange('name')}
                        error={errors.name}
                    />
                    <RenderInput
                        name={'email'}
                        label={'Email'}
                        placeholder={'type email'}
                        value={values.email}
                        onChange={this.onChange('email')}
                        error={errors.email}
                    />
                    <RenderInput
                        name={'password'}
                        label={'Password'}
                        placeholder={'type password'}
                        value={values.password}
                        onChange={this.onChange('password')}
                        error={errors.password}
                    />

                    <button type='submit' disabled={isSubmitDisables}>Отправить</button>
                </form>
            </div>
        );
    }
}


export default App;