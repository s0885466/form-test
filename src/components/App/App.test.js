import React from 'react';
import {shallow, mount, render} from 'enzyme';
import App from "./App";

describe('Testing snapshop <App/>', () => {
    it('App have rendered correctly', () => {
        const app = shallow(<App/>);
        expect(app).toMatchSnapshot();
    })
});

describe('Testing state types <App/>', () => {
    const app = shallow(<App/>);
    it('values should be an Object', () => {
        expect(app.state().values).toBeObject();
    });
    it('isSubmitDisables = true', () => {
        expect(app.state().isSubmitDisables).toBeTrue();
    })
});

describe('Testing handlers <App/>', () => {
    it('onChange name input', () => {
        const app = shallow(<App/>);
        const event = {target: {value: "spam"}};
        app.instance().onChange('name')(event);
        expect(app.state().values.name).toBe(event.target.value);
        expect(app.state().errors.name).toBe('Name must be more than 6 symbols')
    })
});
