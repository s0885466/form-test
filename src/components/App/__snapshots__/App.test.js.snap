// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing snapshop <App/> App have rendered correctly 1`] = `ShallowWrapper {}`;
